services/supervisor/src/api.rs

```diff
@@ -12,6 +12,578 @@
+//!
+//! ```
+//! use exonum::blockchain::ConsensusConfig;
+//! use exonum_rust_runtime::ServiceFactory;
+//! use exonum_supervisor::Supervisor;
+//! use exonum_testkit::{ApiKind, TestKitBuilder};
+//!
+//! # fn main() -> Result<(), failure::Error> {
+//! let mut testkit = TestKitBuilder::validator()
+//!     .with_rust_service(Supervisor)
+//!     .with_artifact(Supervisor.artifact_id())
+//!     .with_instance(Supervisor::simple())
+//!     .create();
+//!
+//! let consensus_config: ConsensusConfig = testkit
+//!     .api()
+//!     .public(ApiKind::Service("supervisor"))
+//!     .get("consensus-config")?;
+//!
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! ## Obtaining Pending Configuration Proposal
+//!
+//! | Property    | Value |
+//! |-------------|-------|
+//! | Path        | `/api/supervisor/v1/config-proposal` |
+//! | Method      | GET   |
+//! | Query type  | - |
+//! | Return type | `Option<[ConfigProposalWithHash]>` |
+//!
+//! Returns the configuration proposal which is currently pending. Returns `None` if there is no
+//! pending configuration at the moment.
+//!
+//! [ConfigProposalWithHash]: ../struct.ConfigProposalWithHash.html
+//!
+//! ```
+//! use exonum_rust_runtime::ServiceFactory;
+//! use exonum_supervisor::{ConfigProposalWithHash, Supervisor};
+//! use exonum_testkit::{ApiKind, TestKitBuilder};
+//!
+//! # fn main() -> Result<(), failure::Error> {
+//! let mut testkit = TestKitBuilder::validator()
+//!     .with_rust_service(Supervisor)
+//!     .with_artifact(Supervisor.artifact_id())
+//!     .with_instance(Supervisor::simple())
+//!     .create();
+//!
+//! let pending_proposal: Option<ConfigProposalWithHash> = testkit
+//!     .api()
+//!     .public(ApiKind::Service("supervisor"))
+//!     .get("config-proposal")?;
+//!
+//! // Will be none, since we did not send a proposal.
+//! assert!(pending_proposal.is_none());
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! # Private API
+//!
+//! ## Request to Deploy an Artifact
+//!
+//! | Property    | Value |
+//! |-------------|-------|
+//! | Path        | `/api/supervisor/v1/deploy-artifact` |
+//! | Method      | POST   |
+//! | Query type  | [`DeployRequest`] |
+//! | Return type | [`Hash`] |
+//!
+//! Requests the deployment of a certain artifact.
+//!
+//! Depending on the supervisor operating mode, it may be required to send such a request to
+//! majority of nodes (in "decentralized" mode), or one request will be enough (in "simple" mode).
+//!
+//! After receiving a deployment request, supervisor creates a corresponding transaction, signs it
+//! with node's keys and broadcasts the transaction within the network. The hash of broadcast
+//! transaction is returned from the endpoint.
+//!
+//! For more details on deploy requests, see [crate documentation](../index.html).
+//!
+//! **Warning:** `DeployRequest` structure should be serialized using corresponding protobuf message,
```

Comment:
Off-topic: that `DeployRequest` is serialized into a hex string is unclear from the code sample below. IMO, a better way would be to have a wrapper type like

```rust
pub struct HexString<T>(pub T);

impl<T: BinaryValue> Serialize for HexString<T> { /* ... */ }

// used like
let response = testkit
    .api()
    .private(ApiKind::Service("supervisor"))
    .query(&HexString(deploy_request))
    .post("deploy-artifact")?;
```

This would have the added benefit that `DeployRequest` could have more "sane" serde serialization in other contexts (e.g., if it is returned from the API, or if somebody decides to serialize it in other format, such as `bincode`).

Also, how is a hex string serialized into query params? It could be provided as is for the request body, but as a query it would require the param name. If it *is* serialized as a request body, it makes sense to replace "query type" -> "body type" in the table above. (And maybe update docs in testkit.)

Reply:
I changed `query` to `body`, and added the notes to the `testkit` docs.

However, I did not introduce `HexString` here, since it's a docs-only PR, and I think it's better to make these changes separately.

Reply:
Sure, that's why I prefaced the comment as off-topic.